src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -285,9 +316,33 @@ public synchronized IndexShard createShard(int sShardId, boolean primary) {
         Injector shardInjector = null;
         try {
 
+            
+
             ShardPath path = ShardPath.loadShardPath(logger, nodeEnv, shardId, indexSettings);
             if (path == null) {
-                path = ShardPath.selectNewPathForShard(nodeEnv, shardId, indexSettings);
+
+                long totFreeSpace = 0;
+                for (NodeEnvironment.NodePath nodePath : nodeEnv.nodePaths()) {
+                    totFreeSpace += nodePath.fileStore.getUsableSpace();
+                }
+
+                // Very rough heurisic of how much disk space we expect the shard will use over its lifetime, the max of current average
+                // shard size across the cluster and 5% of the total available free space on this node:
+                long estShardSizeInBytes = Math.max(avgShardSizeInBytes, (long) (totFreeSpace/20.0));
+
+                // Just collate predicted disk usage on each path.data:
+                Map<Path,Long> estReserveBytes = new HashMap<>();
+                for(Tuple<IndexShard,Injector> shardInjectorTuple : shards.values()) {
+                    IndexShard shard = shardInjectorTuple.v1();
+                    // Remove indices/<index>/<shardID> subdirs from the statePath to get back to the path.data:
+                    Path nodeDataPath = shard.shardPath().getShardStatePath().getParent().getParent().getParent();
```

Comment:
This seems error prone if the structure could change in the future? I dont know a better way though..